### PR TITLE
3235

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#3235 Can't view the table
- Removed typeface calculation and replaced with IDE fonts and fallback

CHS OS:
image

#3248 Some autoformatting annoyances
- Added the requested options
- Range formatter should not use 'exact' version directly, only after it figures out correct range.

Hardened exception handling in the package and function index

### CODE REVIEW

**Comment by MikhailArkhipov** [comment]
@AlexanderSher, @int19h  - need two reviews

**Comment by int19h** [comment]
👍

File: src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs [hunk 1]
```diff
@@ -65,36 +64,43 @@ internal sealed class VisualGrid : FrameworkElement {
...
-                TextElement.FontSizeProperty.AddOwner(
-                        typeof(VisualGrid));
+        public static readonly DependencyProperty FontSizeProperty = TextElement.FontSizeProperty.AddOwner(typeof(VisualGrid));
 
         [TypeConverter(typeof(FontSizeConverter))]
         public double FontSize {
             get { return (double)GetValue(FontSizeProperty); }
-            set { SetValue(FontSizeProperty, value); }
+            set {
+                _typeFace = null;
```
Comment (AlexanderSher) [inline src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs]: DependencyProperty can be changed by WPF using only its metadata without direct access to the property. `DependencyProperty.AddOwner` has overload that allows to specify `FrameworkPropertyMetadata` instance and add custom `PropertyChangedCallback`

File: src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs [hunk 2]
```diff
@@ -65,36 +64,43 @@ internal sealed class VisualGrid : FrameworkElement {
...
+                SetValue(FontWeightProperty, value);
+            }
         }
 
         private Typeface _typeFace;
         private Typeface Typeface {
             get {
                 if (_typeFace == null) {
-                    _typeFace = ChooseTypeface();
+                    if (FontFamily != null && FontSize > 0) {
```
Comment (AlexanderSher) [inline src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs]: Should `Typeface` be a `DependencyProperty` as well?

Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs]: Normally it is not exposed AFACT. Controls have FontSize/Weight/Family/Stretch, never seen Typeface though. I guess this is b/s 4 former ones are defining it and not a single property.

File: src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/VisualGrid/VisualGrid.cs]: OK

File: src/R/Support/Impl/Help/Packages/PackageIndex.cs [hunk 3]
```diff
@@ -232,7 +234,7 @@ public sealed class PackageIndex : IPackageIndex {
                     var installedPackages = await GetInstalledPackagesAsync();
                     var packagesNotInIndex = installedPackages.Where(p => packageNames.Contains(p.Package));
                     info = await AddPackagesToIndexAsync(packagesNotInIndex);
-                } catch (RHostDisconnectedException) { }
+                } catch (RException) { }
```
Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageIndex.cs]: `RHostDisconnectedException` isn't derived from `RException`. Also, `GetFunctionNamesAsync` already handles `REvaluationException`

File: src/R/Support/Impl/Help/Packages/PackageInfo.cs [hunk 4]
```diff
@@ -80,7 +80,7 @@ internal sealed class PackageInfo : NamedItemInfo, IPackageInfo {
                 try {
                     var result = await _host.Session.PackagesFunctionsNamesAsync(Name, REvaluationKind.BaseEnv);
                     functions = result.Children<JValue>().Select(v => (string)v.Value);
-                } catch (TaskCanceledException) { } catch (REvaluationException) { }
+                } catch (RException) { }
```
Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: `} catch (RHostDisconnectedException) { } catch (REvaluationException) {`

File: src/R/Support/Impl/Help/Packages/PackageIndex.cs [hunk 5]
```diff
@@ -260,7 +262,7 @@ public sealed class PackageIndex : IPackageIndex {
             try {
                 await _host.StartSessionAsync();
                 return _host.LoadedPackageNames;
-            } catch (OperationCanceledException) { }
+            } catch (RException) { }
```
Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageIndex.cs]: I think `StartSessionAsync` can't throw `RException`, but it can throw `RHostDisconnectedException`

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageIndex.cs]: You are right. But there is also a number of places when we call things on the session and yet not handle disconnected. Let me look there as well.

File: src/R/Support/Impl/Help/Packages/PackageInfo.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: In most cases in code I see use of `RException` though

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: `catch (RException) { } catch (OperationCanceledException)` should cover most cases I think

Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: We should be careful with it. `catch (RException) { } catch (OperationCanceledException) { }` will swallow all exceptions that we expect to get from `RSession`. 

I think we should rather make it a separate PR.

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: If code already swallows `RException` it should be handling `OperationCanceledException`. In any case, the code is inconsistent - some handle both and some just `RException`. I think it is the old code that we've missed after adding remote. We don't want to crash VS on canceled, right?

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: OK, I pushed related files, will open PR on the rest

Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: No, but there can be side effects:
- `OperationCanceledException` may be required by tasks. For example:
```cs
var cts = new System.Threading.CancellationTokenSource();
var t = Task.Run(async () => {
    while (true) {
        await Task.Delay(1000, cts.Token);
    }
});
```
Task `t` will end when `cts.Cancel()` is called. If we wrap `await Task.Delay(1000, cts.Token);`:
```cs
var cts = new System.Threading.CancellationTokenSource();
var t = Task.Run(async () => {
    while (true) {
        try {
            await Task.Delay(1000, cts.Token);
        } catch (OperationCanceledException) { }
    }
});
```
task will be running forever.

- `RHostDisconnectedException` is captured on the ViewModel layer to show error messages. Swallowing it on the lower levels will break this code.

File: src/R/Support/Impl/Help/Packages/PackageInfo.cs [hunk 6]
```diff
@@ -77,10 +77,8 @@ internal sealed class PackageInfo : NamedItemInfo, IPackageInfo {
         private async Task<IEnumerable<string>> GetFunctionNamesAsync() {
             var functions = TryRestoreFromCache();
             if (functions == null || !functions.Any()) {
-                try {
-                    var result = await _host.Session.PackagesFunctionsNamesAsync(Name, REvaluationKind.BaseEnv);
-                    functions = result.Children<JValue>().Select(v => (string)v.Value);
-                } catch (TaskCanceledException) { } catch (REvaluationException) { }
+                var result = await _host.Session.PackagesFunctionsNamesAsync(Name, REvaluationKind.BaseEnv);
+                functions = result.Children<JValue>().Select(v => (string)v.Value);
```
Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: Previous implementation allowed to proceed with next package in case of error. With this change, `LoadRemainingPackagesFunctions` will stop on the first error. Is it intentional?

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: Probably should skip package on `RException` but skip all on `OperationCanceledException`

Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: Agreed

File: src/R/Support/Impl/Help/Packages/PackageInfo.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: Yes, I just mean that take `LoadWorkspaceCommand` - it handles both

```
        private async Task LoadWorkspace(IRSession session, string file) {
            try {
                await session.LoadWorkspaceAsync(file);
            } catch (RException ex) {
                var message = string.Format(CultureInfo.CurrentCulture, Resources.LoadWorkspaceFailedMessageFormat, file, ex.Message);
                VsAppShell.Current.ShowErrorMessage(message);
            } catch (OperationCanceledException) {
            }
        }
```

but in `WorkingDirectoryCommand` `SetDirectory` does not, only handles `RException`

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: `DeleteAllVariablesCommand` shouldn't be crashing if host disconnects either

**Comment by AlexanderSher** [comment]
👍 
@int19h , do you want to take a look on the PR after all the changed?

File: src/R/Support/Impl/Help/Packages/PackageInfo.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: And it won't. One of the reasons why `RHostDisconnectedException` is derived from `OperationCanceledException` is because `DoNotWait` automatically swallows task cancellation, and so it will swallow `RHostDisconnectedException`. So until we don't expect `rm(list = ls(all = TRUE))` to fail with `REvaluationException`, `DeleteAllVariablesCommand` is fine.

Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: Same thing with `WorkingDirectoryCommand` - `.SilenceException<RException>()` is enough.

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: OK. but there is more that only `catch(RException)`

`AD7Property.SetValueAsStringWithError` - it is COM.
`RExecutionTracer`
`GridDataSource`
`REnvironmentProvider`

and several more

Comment (AlexanderSher) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: `AD7Property.SetValueAsStringWithError` - executed inside `ThreadHelper.JoinableTaskFactory`, which handles task cancellation
`GridDataSource` doesn't catch `RHostDisconnectedException` because calling code relies on it being thrown so that awaiter continuation doesn't execute.
`REnvironmentProvider` catches both

Generally speaking, I'd rather not change anything about handling `RHostDisconnectedException` because:
- for async code, `OperationCanceledException` is a normal thing and doesn't lead to a crash. 
- `RHostDisconnectedException` may be handled up in the stack separately from the `RException` or intentionally not handled, so we can easily break existing functionality
- I don't remember seeing VS crashed because of `RHostDisconnectedException`

Comment (MikhailArkhipov) [inline src/R/Support/Impl/Help/Packages/PackageInfo.cs]: OK ;-) But they would be hard to repro, it needs to disconnect right when code is executing. Which is usually not a very long time.

**Comment by MikhailArkhipov** [comment]
I am going to close this one and merge into Release/1_0 as RC2 is now shipped.

**Comment by AlexanderSher** [comment]
Ok
